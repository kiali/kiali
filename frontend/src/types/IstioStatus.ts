import { t } from 'utils/I18nUtils';

export enum Status {
  Healthy = 'Healthy',
  Unhealthy = 'Unhealthy',
  Unreachable = 'Unreachable',
  NotFound = 'NotFound',
  NotReady = 'NotReady'
}

export const statusMsg = {
  [Status.Healthy]: t('Healthy'),
  [Status.NotFound]: t('Not found'),
  [Status.NotReady]: t('Not ready'),
  [Status.Unhealthy]: t('Not healthy'),
  [Status.Unreachable]: t('Unreachable')
};

export const statusSeverity: Record<Status, number> = {
  [Status.Unhealthy]: 5,
  [Status.Unreachable]: 4,
  [Status.NotReady]: 3,
  [Status.NotFound]: 2,
  [Status.Healthy]: 1
};

export interface ComponentStatus {
  cluster: string;
  is_core: boolean;
  name: string;
  status: Status;
}

export interface IstiodResourceThresholds {
  cpu: number;
  memory: number;
}
