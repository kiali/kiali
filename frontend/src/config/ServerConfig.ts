import _ from 'lodash';
import { ServerConfig } from '../types/ServerConfig';
import { parseHealthConfig } from './HealthConfig';
import { MeshCluster } from '../types/Mesh';

export type Durations = { [key: number]: string };

export type ComputedServerConfig = ServerConfig & {
  durations: Durations;
};

function getHomeCluster(cfg: ServerConfig): MeshCluster | undefined {
  return Object.values(cfg.clusters).find(cluster => cluster.isKialiHome);
}

export const humanDurations = (cfg: ComputedServerConfig, prefix?: string, suffix?: string): Durations =>
  _.mapValues(cfg.durations, v => _.reject([prefix, v, suffix], _.isEmpty).join(' '));

const toDurations = (tupleArray: [number, string][]): Durations => {
  const obj: Duration = {};
  tupleArray.forEach(tuple => {
    obj[tuple[0]] = tuple[1];
  });
  return obj;
};

const durationsTuples: [number, string][] = [
  [60, '1m'],
  [120, '2m'],
  [300, '5m'],
  [600, '10m'],
  [1800, '30m'],
  [3600, '1h'],
  [10800, '3h'],
  [21600, '6h'],
  [43200, '12h'],
  [86400, '1d'],
  [604800, '7d'],
  [2592000, '30d']
];

const computeValidDurations = (cfg: ComputedServerConfig): void => {
  const tsdbRetention = cfg.prometheus.storageTsdbRetention;
  const scrapeInterval = cfg.prometheus.globalScrapeInterval;
  let filtered = durationsTuples.filter(
    d => (!tsdbRetention || d[0] <= tsdbRetention!) && (!scrapeInterval || d[0] >= scrapeInterval * 2)
  );
  // Make sure we keep at least one item, even if it's silly
  if (filtered.length === 0) {
    filtered = [durationsTuples[0]];
  }
  cfg.durations = toDurations(filtered);
};

// Set some reasonable defaults. Initial values should be valid for fields
// than may not be providedby/set on the server.
const defaultServerConfig: ComputedServerConfig = {
  accessibleNamespaces: [],
  ambientEnabled: false,
  authStrategy: '',
  clusters: {},
  durations: {},
  gatewayAPIClasses: [],
  gatewayAPIEnabled: false,
  logLevel: '',
  healthConfig: {
    rate: []
  },
  deployment: {
    viewOnlyMode: false
  },
  installationTag: 'Kiali Console',
  istioAnnotations: {
    ambientAnnotation: 'ambient.istio.io/redirection',
    ambientAnnotationEnabled: 'enabled',
    istioInjectionAnnotation: 'sidecar.istio.io/inject'
  },
  istioCanaryRevision: {
    current: '',
    upgrade: ''
  },
  istioIdentityDomain: 'svc.cluster.local',
  istioNamespace: 'istio-system',
  istioLabels: {
    ambientNamespaceLabel: 'istio.io/dataplane-mode',
    ambientNamespaceLabelValue: 'ambient',
    ambientWaypointLabel: 'gateway.istio.io/managed',
    ambientWaypointLabelValue: 'istio.io-mesh-controller',
    appLabelName: 'app',
    egressGatewayLabel: 'istio=egressgateway',
    ingressGatewayLabel: 'istio=ingressgateway',
    injectionLabelName: 'istio-injection',
    injectionLabelRev: 'istio.io/rev',
    versionLabelName: 'version'
  },
  kialiFeatureFlags: {
    certificatesInformationIndicators: {
      enabled: true
    },
    disabledFeatures: [],
    istioInjectionAction: true,
    istioAnnotationAction: true,
    istioUpgradeAction: false,
    uiDefaults: {
      graph: {
        findOptions: [],
        hideOptions: [],
        impl: 'cy',
        settings: {
          fontLabel: 13,
          minFontBadge: 7,
          minFontLabel: 10
        },
        traffic: {
          grpc: 'requests',
          http: 'requests',
          tcp: 'sent'
        }
      },
      i18n: {
        language: 'en',
        showSelector: false
      },
      list: {
        includeHealth: true,
        includeIstioResources: true,
        includeValidations: true,
        showIncludeToggles: false
      },
      mesh: {
        findOptions: [],
        hideOptions: [],
        impl: 'classic'
      }
    }
  },
  prometheus: {
    globalScrapeInterval: 15,
    storageTsdbRetention: 21600
  }
};

// Overwritten with real server config on user login. Also used for tests.
let serverConfig = defaultServerConfig;
computeValidDurations(serverConfig);
export { serverConfig };

let homeCluster = getHomeCluster(serverConfig);
let isMultiCluster = isMC();
export { homeCluster, isMultiCluster };

export const toValidDuration = (duration: number): number => {
  // Check if valid
  if (serverConfig.durations[duration]) {
    return duration;
  }
  // Get closest duration
  const validDurations = durationsTuples.filter(d => serverConfig.durations[d[0]]);
  for (let i = validDurations.length - 1; i > 0; i--) {
    if (duration > durationsTuples[i][0]) {
      return validDurations[i][0];
    }
  }
  return validDurations[0][0];
};

export const setServerConfig = (cfg: ServerConfig): void => {
  serverConfig = {
    ...defaultServerConfig,
    ...cfg
  };

  serverConfig.healthConfig = cfg.healthConfig ? parseHealthConfig(cfg.healthConfig) : serverConfig.healthConfig;
  computeValidDurations(serverConfig);

  homeCluster = getHomeCluster(serverConfig);
  isMultiCluster = isMC();
};

export const isIstioNamespace = (namespace: string): boolean => {
  if (namespace === serverConfig.istioNamespace) {
    return true;
  }
  return false;
};

export const isHomeCluster = (cluster: string): boolean => {
  return !isMultiCluster || cluster === homeCluster?.name;
};

// Return true if the cluster is configured for this Kiali instance
export const isConfiguredCluster = (cluster: string): boolean => {
  return Object.keys(serverConfig.clusters).includes(cluster);
};

function isMC(): boolean {
  // If there is only one cluster, it is not a multi-cluster deployment.
  // If there are multiple clusters but only one is accessible, it is not a multi-cluster deployment.
  return (
    Object.keys(serverConfig.clusters).length > 1 &&
    Object.values(serverConfig.clusters).filter(c => c.accessible).length > 1
  );
}
