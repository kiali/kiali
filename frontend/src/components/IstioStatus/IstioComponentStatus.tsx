import * as React from 'react';
import { ComponentStatus, Status, statusMsg } from '../../types/IstioStatus';
import { PFColors } from '../Pf/PfColors';
import {
  CheckCircleIcon,
  ExclamationCircleIcon,
  ExclamationTriangleIcon,
  MinusCircleIcon
} from '@patternfly/react-icons';
import { Split, SplitItem } from '@patternfly/react-core';
import { IconProps, createIcon } from 'config/KialiIcon';
import { kialiStyle } from 'styles/StyleUtils';
import { useKialiTranslation } from 'utils/I18nUtils';

type Props = {
  componentStatus: ComponentStatus;
};

const ErrorCoreComponent: IconProps = {
  color: PFColors.Danger,
  icon: ExclamationCircleIcon
};

const ErrorAddonComponent: IconProps = {
  color: PFColors.Warning,
  icon: ExclamationTriangleIcon
};

const NotReadyComponent: IconProps = {
  color: PFColors.Info,
  icon: MinusCircleIcon
};

const SuccessComponent: IconProps = {
  color: PFColors.Success,
  icon: CheckCircleIcon
};

// Mapping Valid-Core to Icon representation.
const validToIcon: { [valid: string]: IconProps } = {
  'false-false': ErrorAddonComponent,
  'false-true': ErrorCoreComponent,
  'true-false': SuccessComponent,
  'true-true': SuccessComponent
};

const splitItemStyle = kialiStyle({
  textAlign: 'left'
});

export const IstioComponentStatus: React.FC<Props> = (props: Props) => {
  const { t } = useKialiTranslation();

  const renderIcon = (status: Status, isCore: boolean): React.ReactNode => {
    let compIcon = validToIcon[`${status === Status.Healthy}-${isCore}`];

    if (status === Status.NotReady) {
      compIcon = NotReadyComponent;
    }

    compIcon.className = kialiStyle({
      marginTop: '0.25rem'
    });

    return createIcon(compIcon);
  };

  const renderCells = (): React.ReactNode => {
    const comp = props.componentStatus;

    return [
      <Split key={`cell-status-icon-${comp.name}`} hasGutter={true} className={splitItemStyle}>
        <SplitItem>{renderIcon(props.componentStatus.status, props.componentStatus.isCore)}</SplitItem>
        <SplitItem isFilled={true}>{comp.name}</SplitItem>
        <SplitItem>{t(statusMsg[comp.status])}</SplitItem>
      </Split>
    ];
  };

  return <>{renderCells()}</>;
};
