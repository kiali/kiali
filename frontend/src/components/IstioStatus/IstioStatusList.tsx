import * as React from 'react';
import { List } from '@patternfly/react-core';
import { ComponentStatus, Status } from '../../types/IstioStatus';
import { IstioComponentStatus } from './IstioComponentStatus';
import { kialiStyle } from 'styles/StyleUtils';
import { useKialiTranslation } from 'utils/I18nUtils';

type Props = {
  cluster: string;
  status: ComponentStatus[];
};

const listStyle = kialiStyle({
  paddingLeft: 0,
  marginTop: 0,
  marginLeft: 0
});

export const IstioStatusList: React.FC<Props> = (props: Props) => {
  const { t } = useKialiTranslation();

  const nonhealthyComponents = (): ComponentStatus[] => {
    return props.status.filter((c: ComponentStatus) => c.status !== Status.Healthy);
  };

  const coreComponentsStatus = (): ComponentStatus[] => {
    return nonhealthyComponents().filter((s: ComponentStatus) => s.is_core);
  };

  const addonComponentsStatus = (): ComponentStatus[] => {
    return nonhealthyComponents().filter((s: ComponentStatus) => !s.is_core);
  };

  const renderComponentList = (): React.ReactNode => {
    const groups = {
      core: coreComponentsStatus,
      addon: addonComponentsStatus
    };

    return ['core', 'addon'].map((group: string) => {
      return (
        <React.Fragment key={`status-${group}`}>
          {groups[group]().map((status: ComponentStatus) => {
            return <IstioComponentStatus key={`status-${group}-${status.name}`} componentStatus={status} />;
          })}
        </React.Fragment>
      );
    });
  };

  return (
    <List id="istio-status" aria-label={t('Istio Component List')} className={listStyle}>
      {renderComponentList()}
    </List>
  );
};
