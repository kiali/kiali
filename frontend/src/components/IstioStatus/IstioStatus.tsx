import * as React from 'react';
import { SVGIconProps } from '@patternfly/react-icons/dist/js/createIcon';
import * as API from '../../services/Api';
import * as AlertUtils from '../../utils/AlertUtils';
import { TimeInMilliseconds } from '../../types/Common';
import { ComponentStatus, Status } from '../../types/IstioStatus';
import { MessageType } from '../../types/MessageCenter';
import { Namespace } from '../../types/Namespace';
import { KialiAppState } from '../../store/Store';
import { istioStatusSelector, namespaceItemsSelector } from '../../store/Selectors';
import { bindActionCreators } from 'redux';
import { IstioStatusActions } from '../../actions/IstioStatusActions';
import { connect } from 'react-redux';
import { Tooltip, TooltipPosition } from '@patternfly/react-core';
import { IstioStatusList } from './IstioStatusList';
import { PFColors } from '../Pf/PfColors';
import { ResourcesFullIcon } from '@patternfly/react-icons';
import { KialiDispatch } from 'types/Redux';
import { NamespaceThunkActions } from '../../actions/NamespaceThunkActions';
import { connectRefresh } from '../Refresh/connectRefresh';
import { kialiStyle } from 'styles/StyleUtils';
import { IconProps, createIcon } from 'config/KialiIcon';
import { Link } from 'react-router-dom-v5-compat';
import { useKialiTranslation } from 'utils/I18nUtils';
import { MASTHEAD } from 'components/Nav/Masthead/Masthead';
import { isControlPlaneAccessible } from '../../utils/MeshUtils';

type ReduxStateProps = {
  namespaces?: Namespace[];
  status: ComponentStatus[];
};

type ReduxDispatchProps = {
  refreshNamespaces: () => void;
  setIstioStatus: (istioStatus: ComponentStatus[]) => void;
};

type StatusIcons = {
  ErrorIcon?: React.ComponentClass<SVGIconProps>;
  HealthyIcon?: React.ComponentClass<SVGIconProps>;
  InfoIcon?: React.ComponentClass<SVGIconProps>;
  WarningIcon?: React.ComponentClass<SVGIconProps>;
};

type Props = ReduxStateProps &
  ReduxDispatchProps & {
    cluster?: string;
    icons?: StatusIcons;
    lastRefreshAt: TimeInMilliseconds;
    location?: string;
  };

const ValidToColor = {
  'true-true-true': PFColors.Danger,
  'true-true-false': PFColors.Danger,
  'true-false-true': PFColors.Danger,
  'true-false-false': PFColors.Danger,
  'false-true-true': PFColors.Warning,
  'false-true-false': PFColors.Warning,
  'false-false-true': PFColors.Info,
  'false-false-false': PFColors.Success
};

const defaultIcons = {
  ErrorIcon: ResourcesFullIcon,
  HealthyIcon: ResourcesFullIcon,
  InfoIcon: ResourcesFullIcon,
  WarningIcon: ResourcesFullIcon
};

const iconStyle = kialiStyle({
  marginLeft: '2rem',
  fontSize: '1rem'
});

export const meshLinkStyle = kialiStyle({
  display: 'flex',
  justifyContent: 'center',
  marginTop: '0.75rem',
  $nest: {
    '& > span': {
      marginRight: '0.5rem'
    }
  }
});

export const IstioStatusComponent: React.FC<Props> = (props: Props) => {
  const { t } = useKialiTranslation();

  const { cluster, namespaces, setIstioStatus, refreshNamespaces, lastRefreshAt } = props;

  React.useEffect(() => {
    refreshNamespaces();
  }, [refreshNamespaces]);

  const fetchStatus = React.useCallback((): void => {
    API.getIstioStatus(cluster)
      .then(response => {
        setIstioStatus(response.data);
      })
      .catch(error => {
        // User without namespaces can't have access to mTLS information. Reduce severity to info.
        const informative = namespaces && namespaces.length < 1;

        if (informative) {
          AlertUtils.addError(t('Istio deployment status disabled.'), error, 'default', MessageType.INFO);
        } else {
          AlertUtils.addError(t('Error fetching Istio deployment status.'), error, 'default', MessageType.ERROR);
        }
      });
  }, [cluster, namespaces, setIstioStatus, t]);

  React.useEffect(() => {
    fetchStatus();
  }, [lastRefreshAt, fetchStatus]);

  const tooltipContent = (): React.ReactNode => {
    return (
      <>
        <IstioStatusList status={props.status} />

        {!props.location?.endsWith('/mesh') && isControlPlaneAccessible() && (
          <div className={meshLinkStyle}>
            <span>{t('More info at')}</span>
            <Link to="/mesh">{t('Mesh page')}</Link>
          </div>
        )}
      </>
    );
  };

  const tooltipColor = (): string => {
    let coreUnhealthy = false;
    let addonUnhealthy = false;
    let notReady = false;

    Object.keys(props.status ?? {}).forEach((compKey: string) => {
      const { status, is_core } = props.status[compKey];
      const isNotReady: boolean = status === Status.NotReady;
      const isUnhealthy: boolean = status !== Status.Healthy && !isNotReady;

      if (is_core) {
        coreUnhealthy = coreUnhealthy || isUnhealthy;
      } else {
        addonUnhealthy = addonUnhealthy || isUnhealthy;
      }

      notReady = notReady || isNotReady;
    });

    return ValidToColor[`${coreUnhealthy}-${addonUnhealthy}-${notReady}`];
  };

  const healthyComponents = (): boolean => {
    return props.status.reduce((healthy: boolean, compStatus: ComponentStatus) => {
      return healthy && compStatus.status === Status.Healthy;
    }, true);
  };

  if (!healthyComponents()) {
    const icons = props.icons ? { ...defaultIcons, ...props.icons } : defaultIcons;
    const iconColor = tooltipColor();
    let icon = ResourcesFullIcon;
    let dataTest = 'istio-status';

    if (iconColor === PFColors.Danger) {
      icon = icons.ErrorIcon;
      dataTest = `${dataTest}-danger`;
    } else if (iconColor === PFColors.Warning) {
      icon = icons.WarningIcon;
      dataTest = `${dataTest}-warning`;
    } else if (iconColor === PFColors.Info) {
      icon = icons.InfoIcon;
      dataTest = `${dataTest}-info`;
    } else if (iconColor === PFColors.Success) {
      icon = icons.HealthyIcon;
      dataTest = `${dataTest}-success`;
    }

    const iconProps: IconProps = {
      className: iconStyle,
      dataTest: dataTest
    };

    const tooltipPosition = props.location === MASTHEAD ? TooltipPosition.bottom : TooltipPosition.top;

    return (
      <Tooltip position={tooltipPosition} enableFlip={true} content={tooltipContent()} maxWidth="25rem">
        {createIcon(iconProps, icon, iconColor)}
      </Tooltip>
    );
  }

  return null;
};

const mapStateToProps = (state: KialiAppState): ReduxStateProps => ({
  status: istioStatusSelector(state),
  namespaces: namespaceItemsSelector(state)
});

const mapDispatchToProps = (dispatch: KialiDispatch): ReduxDispatchProps => ({
  setIstioStatus: bindActionCreators(IstioStatusActions.setinfo, dispatch),
  refreshNamespaces: () => {
    dispatch(NamespaceThunkActions.fetchNamespacesIfNeeded());
  }
});

export const IstioStatus = connectRefresh(connect(mapStateToProps, mapDispatchToProps)(IstioStatusComponent));
