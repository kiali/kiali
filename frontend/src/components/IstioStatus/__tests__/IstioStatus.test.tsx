import * as React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import { ComponentStatus, Status } from '../../../types/IstioStatus';
import { IstioStatusComponent } from '../IstioStatus';
import { shallowToJson } from 'enzyme-to-json';
import { CLUSTER_DEFAULT } from '../../../types/Graph';
import { serverConfig } from '../../../config';
import { setServerConfig } from '../../../config/ServerConfig';

const mockIcon = (componentList: ComponentStatus[]): ShallowWrapper => {
  return shallow(
    <IstioStatusComponent
      statusMap={{ Kubernetes: componentList }}
      lastRefreshAt={848152}
      namespaces={[
        { name: 'bookinfo', cluster: CLUSTER_DEFAULT },
        { name: 'istio-system', cluster: CLUSTER_DEFAULT }
      ]}
      setIstioStatus={jest.fn()}
      refreshNamespaces={jest.fn()}
    />
  );
};

const testSnapshot = (wrapper: any): void => {
  expect(shallowToJson(wrapper)).toBeDefined();
  expect(shallowToJson(wrapper)).toMatchSnapshot();
};

const testTooltip = (wrapper: any): void => {
  expect(wrapper.name()).toEqual('Tooltip');
  expect(wrapper.props().position).toEqual('top');
  expect(wrapper.props().enableFlip).toEqual(true);
  expect(wrapper.children.length).toEqual(1);
};

const testIcon = (wrapper: any, dataTest: string, iconName: string): void => {
  const clusterIcon = wrapper.childAt(0);
  expect(clusterIcon).toBeDefined();
  const icon = clusterIcon.childAt(1);
  expect(icon).toBeDefined();
  expect(icon.props()['data-test']).toEqual(dataTest);
  expect(icon.childAt(0).name()).toEqual(iconName);
};

jest.mock('../../../utils/MeshUtils', () => ({
  isControlPlaneAccessible: () => true
}));

describe('When core component has a problem', () => {
  beforeEach(() => {
    serverConfig.clusters = {
      CLUSTER_DEFAULT: {
        accessible: true,
        apiEndpoint: '',
        isKialiHome: true,
        kialiInstances: [],
        name: CLUSTER_DEFAULT,
        secretName: 'test-secret'
      }
    };
    setServerConfig(serverConfig);
  });

  it('the Icon shows is displayed in Red', () => {
    const wrapper = mockIcon([
      {
        cluster: CLUSTER_DEFAULT,
        name: 'grafana',
        status: Status.Healthy,
        is_core: false
      },
      {
        cluster: CLUSTER_DEFAULT,
        name: 'istio-egressgateway',
        status: Status.Unhealthy,
        is_core: true
      }
    ]);

    testSnapshot(wrapper);
    testTooltip(wrapper);
    testIcon(wrapper, 'istio-status-danger', 'ExclamationCircleIcon');
  });
});

describe('When addon component has a problem', () => {
  it('the Icon shows is displayed in orange', () => {
    const wrapper = mockIcon([
      {
        cluster: CLUSTER_DEFAULT,
        name: 'grafana',
        status: Status.Unhealthy,
        is_core: false
      },
      {
        cluster: CLUSTER_DEFAULT,
        name: 'istio-egressgateway',
        status: Status.Healthy,
        is_core: true
      }
    ]);

    testSnapshot(wrapper);
    testTooltip(wrapper);
    testIcon(wrapper, 'istio-status-warning', 'ExclamationTriangleIcon');
  });
});

describe('When both core and addon component have problems', () => {
  describe('any component is in not ready', () => {
    it('the Icon shows is displayed in red', () => {
      const wrapper = mockIcon([
        {
          cluster: CLUSTER_DEFAULT,
          name: 'grafana',
          status: Status.Unhealthy,
          is_core: false
        },
        {
          cluster: CLUSTER_DEFAULT,
          name: 'istio-egressgateway',
          status: Status.Unhealthy,
          is_core: true
        }
      ]);

      testSnapshot(wrapper);
      testTooltip(wrapper);
      testIcon(wrapper, 'istio-status-danger', 'ExclamationCircleIcon');
    });
  });
});

describe('When there are not-ready components', () => {
  describe('mixed with other not healthy components', () => {
    describe('in core', () => {
      it('the Icon is displayed in red', () => {
        const wrapper = mockIcon([
          {
            cluster: CLUSTER_DEFAULT,
            name: 'istio-egressgateway',
            status: Status.Unhealthy,
            is_core: true
          },
          {
            cluster: CLUSTER_DEFAULT,
            name: 'istio-ingressgateway',
            status: Status.NotReady,
            is_core: true
          }
        ]);

        testSnapshot(wrapper);
        testTooltip(wrapper);
        testIcon(wrapper, 'istio-status-danger', 'ExclamationCircleIcon');
      });
    });

    describe('in addons', () => {
      it('the Icon is displayed in orange', () => {
        const wrapper = mockIcon([
          {
            cluster: CLUSTER_DEFAULT,
            name: 'grafana',
            status: Status.Unhealthy,
            is_core: false
          },
          {
            cluster: CLUSTER_DEFAULT,
            name: 'jaeger',
            status: Status.NotReady,
            is_core: false
          }
        ]);

        testSnapshot(wrapper);
        testTooltip(wrapper);
        testIcon(wrapper, 'istio-status-warning', 'ExclamationTriangleIcon');
      });
    });

    describe('in both', () => {
      it('the Icon shows is displayed in red', () => {
        const wrapper = mockIcon([
          {
            cluster: CLUSTER_DEFAULT,
            name: 'grafana',
            status: Status.Unhealthy,
            is_core: false
          },
          {
            cluster: CLUSTER_DEFAULT,
            name: 'jaeger',
            status: Status.NotReady,
            is_core: false
          },
          {
            cluster: CLUSTER_DEFAULT,
            name: 'istio-egressgateway',
            status: Status.Unhealthy,
            is_core: true
          },
          {
            cluster: CLUSTER_DEFAULT,
            name: 'istio-ingressgateway',
            status: Status.NotReady,
            is_core: true
          }
        ]);

        testSnapshot(wrapper);
        testTooltip(wrapper);
        testIcon(wrapper, 'istio-status-danger', 'ExclamationCircleIcon');
      });
    });
  });

  describe('not mixed with other unhealthy components', () => {
    describe('in core', () => {
      it('renders the Icon in blue', () => {
        const wrapper = mockIcon([
          {
            cluster: CLUSTER_DEFAULT,
            name: 'jaeger',
            status: Status.NotReady,
            is_core: false
          }
        ]);

        testSnapshot(wrapper);
        testTooltip(wrapper);
        testIcon(wrapper, 'istio-status-info', 'InfoCircleIcon');
      });
    });

    describe('in addons', () => {
      it('renders the Icon in blue', () => {
        const wrapper = mockIcon([
          {
            cluster: CLUSTER_DEFAULT,
            name: 'istiod',
            status: Status.NotReady,
            is_core: true
          }
        ]);

        testSnapshot(wrapper);
        testTooltip(wrapper);
        testIcon(wrapper, 'istio-status-info', 'InfoCircleIcon');
      });
    });
  });
});

describe('When all components are good', () => {
  it('the Icon shows is displayed in green', () => {
    const wrapper = mockIcon([
      {
        cluster: CLUSTER_DEFAULT,
        name: 'grafana',
        status: Status.Healthy,
        is_core: false
      },
      {
        cluster: CLUSTER_DEFAULT,
        name: 'istio-egressgateway',
        status: Status.Healthy,
        is_core: true
      }
    ]);

    testIcon(wrapper, 'istio-status-success', 'CheckCircleIcon');
  });
});
