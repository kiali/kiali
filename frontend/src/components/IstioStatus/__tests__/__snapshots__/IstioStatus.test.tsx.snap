// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When addon component has a problem the Icon shows is displayed in orange 1`] = `
<Tooltip
  content={
    <React.Fragment>
      <IstioStatusList
        status={
          Array [
            Object {
              "is_core": false,
              "name": "grafana",
              "status": "Unhealthy",
            },
            Object {
              "is_core": true,
              "name": "istio-egressgateway",
              "status": "Healthy",
            },
          ]
        }
      />
    </React.Fragment>
  }
  enableFlip={true}
  maxWidth="25rem"
  position="top"
>
  <Icon
    className="kiali_f1wuoq20 kiali_frkkrth"
    data-test="istio-status-warning"
  >
    <ResourcesFullIcon />
  </Icon>
</Tooltip>
`;

exports[`When both core and addon component have problems any component is in not ready the Icon shows is displayed in red 1`] = `
<Tooltip
  content={
    <React.Fragment>
      <IstioStatusList
        status={
          Array [
            Object {
              "is_core": false,
              "name": "grafana",
              "status": "Unhealthy",
            },
            Object {
              "is_core": true,
              "name": "istio-egressgateway",
              "status": "Unhealthy",
            },
          ]
        }
      />
    </React.Fragment>
  }
  enableFlip={true}
  maxWidth="25rem"
  position="top"
>
  <Icon
    className="kiali_f1wuoq20 kiali_f1h37jb8"
    data-test="istio-status-danger"
  >
    <ResourcesFullIcon />
  </Icon>
</Tooltip>
`;

exports[`When core component has a problem the Icon shows is displayed in Red 1`] = `
<Tooltip
  content={
    <React.Fragment>
      <IstioStatusList
        status={
          Array [
            Object {
              "is_core": false,
              "name": "grafana",
              "status": "Healthy",
            },
            Object {
              "is_core": true,
              "name": "istio-egressgateway",
              "status": "Unhealthy",
            },
          ]
        }
      />
    </React.Fragment>
  }
  enableFlip={true}
  maxWidth="25rem"
  position="top"
>
  <Icon
    className="kiali_f1wuoq20 kiali_f1h37jb8"
    data-test="istio-status-danger"
  >
    <ResourcesFullIcon />
  </Icon>
</Tooltip>
`;

exports[`When there are not-ready components mixed with other not healthy components in addons the Icon is displayed in orange 1`] = `
<Tooltip
  content={
    <React.Fragment>
      <IstioStatusList
        status={
          Array [
            Object {
              "is_core": false,
              "name": "grafana",
              "status": "Unhealthy",
            },
            Object {
              "is_core": false,
              "name": "jaeger",
              "status": "NotReady",
            },
          ]
        }
      />
    </React.Fragment>
  }
  enableFlip={true}
  maxWidth="25rem"
  position="top"
>
  <Icon
    className="kiali_f1wuoq20 kiali_frkkrth"
    data-test="istio-status-warning"
  >
    <ResourcesFullIcon />
  </Icon>
</Tooltip>
`;

exports[`When there are not-ready components mixed with other not healthy components in both the Icon shows is displayed in red 1`] = `
<Tooltip
  content={
    <React.Fragment>
      <IstioStatusList
        status={
          Array [
            Object {
              "is_core": false,
              "name": "grafana",
              "status": "Unhealthy",
            },
            Object {
              "is_core": false,
              "name": "jaeger",
              "status": "NotReady",
            },
            Object {
              "is_core": true,
              "name": "istio-egressgateway",
              "status": "Unhealthy",
            },
            Object {
              "is_core": true,
              "name": "istio-ingressgateway",
              "status": "NotReady",
            },
          ]
        }
      />
    </React.Fragment>
  }
  enableFlip={true}
  maxWidth="25rem"
  position="top"
>
  <Icon
    className="kiali_f1wuoq20 kiali_f1h37jb8"
    data-test="istio-status-danger"
  >
    <ResourcesFullIcon />
  </Icon>
</Tooltip>
`;

exports[`When there are not-ready components mixed with other not healthy components in core the Icon is displayed in red 1`] = `
<Tooltip
  content={
    <React.Fragment>
      <IstioStatusList
        status={
          Array [
            Object {
              "is_core": true,
              "name": "istio-egressgateway",
              "status": "Unhealthy",
            },
            Object {
              "is_core": true,
              "name": "istio-ingressgateway",
              "status": "NotReady",
            },
          ]
        }
      />
    </React.Fragment>
  }
  enableFlip={true}
  maxWidth="25rem"
  position="top"
>
  <Icon
    className="kiali_f1wuoq20 kiali_f1h37jb8"
    data-test="istio-status-danger"
  >
    <ResourcesFullIcon />
  </Icon>
</Tooltip>
`;

exports[`When there are not-ready components not mixed with other unhealthy components in addons renders the Icon in blue 1`] = `
<Tooltip
  content={
    <React.Fragment>
      <IstioStatusList
        status={
          Array [
            Object {
              "is_core": true,
              "name": "istiod",
              "status": "NotReady",
            },
          ]
        }
      />
    </React.Fragment>
  }
  enableFlip={true}
  maxWidth="25rem"
  position="top"
>
  <Icon
    className="kiali_f1wuoq20 kiali_f1elceoh"
    data-test="istio-status-info"
  >
    <ResourcesFullIcon />
  </Icon>
</Tooltip>
`;

exports[`When there are not-ready components not mixed with other unhealthy components in core renders the Icon in blue 1`] = `
<Tooltip
  content={
    <React.Fragment>
      <IstioStatusList
        status={
          Array [
            Object {
              "is_core": false,
              "name": "jaeger",
              "status": "NotReady",
            },
          ]
        }
      />
    </React.Fragment>
  }
  enableFlip={true}
  maxWidth="25rem"
  position="top"
>
  <Icon
    className="kiali_f1wuoq20 kiali_f1elceoh"
    data-test="istio-status-info"
  >
    <ResourcesFullIcon />
  </Icon>
</Tooltip>
`;
